### PR TITLE
ngtsc support for @Pipes!

### PR DESCRIPTION
**Pull request by alxhub**

### CODE REVIEW

**Review by mhevery — DISMISSED** [review]

File: packages/compiler-cli/src/ngtsc/annotations/src/component.ts [hunk 1]
```diff
@@ -82,6 +82,8 @@ export class ComponentDecoratorHandler implements DecoratorHandler<R3ComponentMe
           `Errors parsing template: ${template.errors.map(e => e.toString()).join(', ')}`);
     }
 
+    console.error('compiling component', node.name !.text);
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/annotations/src/component.ts]: Done

File: packages/compiler-cli/src/ngtsc/annotations/src/selector_scope.ts [hunk 2]
```diff
@@ -107,6 +107,7 @@ export class SelectorScopeRegistry {
    * Register the selector of a component or directive with the registry.
    */
   registerSelector(node: ts.Declaration, selector: string): void {
+    console.error('register selector', selector);
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/annotations/src/selector_scope.ts]: Done

File: packages/compiler-cli/src/ngtsc/annotations/src/selector_scope.ts [hunk 3]
```diff
@@ -118,7 +119,12 @@ export class SelectorScopeRegistry {
   /**
    * Register the name of a pipe with the registry.
    */
-  registerPipe(node: ts.Declaration, name: string): void { this._pipeToName.set(node, name); }
+  registerPipe(node: ts.Declaration, name: string): void {
+    console.error('register pipe', name);
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/annotations/src/selector_scope.ts]: Done

File: packages/compiler-cli/src/ngtsc/annotations/src/selector_scope.ts [hunk 4]
```diff
@@ -153,10 +159,23 @@ export class SelectorScopeRegistry {
     // The initial value of ngModuleImportedFrom is 'null' which signifies that the NgModule
     // was not imported from a .d.ts source.
     this.lookupScopes(module !, /* ngModuleImportedFrom */ null).compilation.forEach(ref => {
-      const selector = this.lookupDirectiveSelector(ts.getOriginalNode(ref.node) as ts.Declaration);
+      const node = ts.getOriginalNode(ref.node) as ts.Declaration;
+
+      console.error('processing node', (node as ts.ClassDeclaration).name !.text);
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/annotations/src/selector_scope.ts]: Done

File: packages/compiler-cli/src/ngtsc/annotations/src/selector_scope.ts [hunk 5]
```diff
@@ -153,10 +159,23 @@ export class SelectorScopeRegistry {
...
       // Only directives/components with selectors get added to the scope.
       if (selector != null) {
         directives.set(selector, ref);
+        console.error('is Directive with selector', selector);
+        return;
+      }
+
+      const name = this.lookupPipeName(node);
+      if (name != null) {
+        console.error('is Pipe with name', name);
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/annotations/src/selector_scope.ts]: Done

File: packages/compiler-cli/src/ngtsc/annotations/src/selector_scope.ts [hunk 6]
```diff
@@ -153,10 +159,23 @@ export class SelectorScopeRegistry {
...
+      const node = ts.getOriginalNode(ref.node) as ts.Declaration;
+
+      console.error('processing node', (node as ts.ClassDeclaration).name !.text);
+
+      // Either the node represents a directive or a pipe. Look for both.
+      const selector = this.lookupDirectiveSelector(node);
       // Only directives/components with selectors get added to the scope.
       if (selector != null) {
         directives.set(selector, ref);
+        console.error('is Directive with selector', selector);
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/annotations/src/selector_scope.ts]: Done

File: packages/core/src/render3/jit/pipe.ts [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+import {angularCoreEnv} from './environment';
+import {NG_PIPE_DEF} from './fields';
+import {reflectDependencies} from './util';
+
+export function compilePipe(type: Type<any>, meta: Pipe): void {
+  let def: any = null;
+  Object.defineProperty(type, NG_PIPE_DEF, {
+    get: () => {
+      if (def === null) {
+        const sourceMapUrl = `ng://${type && type.name}/ngPipeDef.js`;
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/core/src/render3/jit/pipe.ts]: Done.

**Comment by mhevery** [comment]
http://test/OCL:203127538:BASE:203127583:1530632222757:c6108879

File: packages/compiler-cli/src/ngtsc/annotations/src/pipe.ts [hunk 8]
```diff
@@ -6,34 +6,76 @@
...
+    if (!pipe.has('name')) {
+      throw new Error(`@Pipe decorator is missing name field`);
+    }
+    const pipeName = staticallyResolve(pipe.get('name') !, this.checker);
+    if (typeof pipeName !== 'string') {
+      throw new Error(`@Pipe.name must be a string`);
+    }
+    this.scopeRegistry.registerPipe(clazz, pipeName);
+
+    let pure = false;
```
Comment (cexbrayat) [inline packages/compiler-cli/src/ngtsc/annotations/src/pipe.ts]: @alxhub Dumb question as I'm probably missing something but shouldn't that be `let pure = true` by default as it is with the current renderer and in Ivy JiT mode?

Comment (cexbrayat) [reply packages/compiler-cli/src/ngtsc/annotations/src/pipe.ts]: I opened PR #24750 in case the pipes should indeed by pure by default.

Comment (mhevery) [reply packages/compiler-cli/src/ngtsc/annotations/src/pipe.ts]: thank you.

**Pull request closed without merging**